src/BinSkim.Rules/SignSecurely.cs

```diff
@@ -0,0 +1,456 @@
+            // TODO. This code could to ensure that retrieved certificate matches
+            // against an allow list. This functionality will require adding
+            // appropriate options to this rule
+#if NOT_DEFINED
+            // #define SHA256_HASH_LEN 32
+            uint cbKeyId = 32;
+            byte[] rgbKeyId = new byte[cbKeyId];
+            string key;
+
+            if (!Native.CryptHashPublicKeyInfo(
+                IntPtr.Zero,
+                Native.CALG_SHA_256,
+                0,
+                Native.X509_ASN_ENCODING,
+                ref certInfo.SubjectPublicKeyInfo,
+                rgbKeyId,
+                ref cbKeyId
+                ))
+            {
+                validationResult = (ValidationResult)Marshal.GetLastWin32Error();
+                return;
+            }
+
+            // Strip hyphens to match output as provided by tools such as certutil.exe 
+            key = BitConverter.ToString(rgbKeyId).Replace('-', ' ');
+
+            // This call to HashSet.Contains is case-insensitive due to initializing the hashset with an appropriate comparer
+            validationResult = (trustedKeys.Contains(key) ? ValidationResult.Valid : ValidationResult.UnknownCertificateKey);
+#endif
+            return true;
+        }
+
+        private void InvokeCloseAction(Native.WINTRUST_DATA winTrustData)
+        {
+            IntPtr pWinTrustData = IntPtr.Zero;
+            Guid action = Native.ActionGenericVerifyV2;
+            winTrustData.StateAction = Native.StateAction.WTD_STATEACTION_CLOSE;
+
+            try
+            {
+                pWinTrustData = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(Native.WINTRUST_DATA)));
+                Marshal.StructureToPtr(winTrustData, pWinTrustData, false);
+
+                Native.WinVerifyTrust(Native.INVALID_HANDLE_VALUE, ref action, pWinTrustData);
+            }
+            finally
+            {
+                if (pWinTrustData != IntPtr.Zero)
+                {
+                    Marshal.FreeHGlobal(pWinTrustData);
+                }
+                Debug.Assert(winTrustData.File == IntPtr.Zero);
+            }
+        }
+
+        private ValidationResult WinVerifyTrustHelper(IntPtr handle, ref Native.WINTRUST_DATA winTrustData)
+        {
+            Guid action = Native.ActionGenericVerifyV2;
+
+            ValidationResult validationResult;
+            IntPtr pWinTrustData = IntPtr.Zero;
+
+            try
+            {
+                pWinTrustData = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(Native.WINTRUST_DATA)));
+                Marshal.StructureToPtr(winTrustData, pWinTrustData, false);
+
+                validationResult = (ValidationResult)Native.WinVerifyTrust(handle, ref action, pWinTrustData);
+
+                winTrustData = (Native.WINTRUST_DATA)Marshal.PtrToStructure(pWinTrustData, typeof(Native.WINTRUST_DATA));
+            }
+            finally
+            {
+                if (pWinTrustData != IntPtr.Zero)
```

Comment:
(And if this has you thinking "Wow, WinVerifyTrust is the worst API design ever", you're not wrong) :P